src/Raven.Server/Documents/Subscriptions/LiveSubscriptionPerformanceCollector.cs

```diff
@@ -0,0 +1,305 @@
+                        // add batch stats to results
+                        var subscriptionItem = results.Find(x => x.TaskId == kvp.Value.Handler.Connection?.SubscriptionId);
+                        if (subscriptionItem != null)
+                        {
+                            var batchPerformance = batchesAggregators.Select(x => x.ToBatchPerformanceLiveStatsWithDetails());
+                            subscriptionItem.BatchPerformance = batchesAggregators.Select(x => x.ToBatchPerformanceLiveStatsWithDetails()).ToArray();
+                        }
+                    }
+                }
+            }
+            
+            return results;
+        }
+
+        protected override async Task StartCollectingStats()
+        {
+            Database.SubscriptionStorage.OnEndConnectionEvent += OnEndConnection;
+            Database.SubscriptionStorage.OnEndBatchEvent += OnEndBatch;
+            Database.SubscriptionStorage.OnTaskAddedEvent += OnSubscriptionTaskAdded;
+            Database.SubscriptionStorage.OnTaskRemovedEvent += OnSubscriptionTaskRemoved;
+
+            try
+            {
+                await RunInLoop();
+            }
+            finally
+            {
+                Database.SubscriptionStorage.OnEndConnectionEvent -= OnEndConnection;
+                Database.SubscriptionStorage.OnEndBatchEvent -= OnEndBatch;
+                Database.SubscriptionStorage.OnTaskAddedEvent -= OnSubscriptionTaskAdded;
+                Database.SubscriptionStorage.OnTaskRemovedEvent -= OnSubscriptionTaskRemoved;
+            }
+        }
+        
+        // Get periodic info for ws
+        protected override List<SubscriptionTaskPerformanceStats> PreparePerformanceStats()
+        {
+            var preparedStats = new List<SubscriptionTaskPerformanceStats>();
+
+            foreach (var kvp in _perSubscriptionConnectionStats)
+            {
+                var connectionPerformance = new List<SubscriptionConnectionPerformanceStats>();
+                
+                var subscriptionAndPerformanceConnectionStatsList = kvp.Value;
+                var subscriptionName = subscriptionAndPerformanceConnectionStatsList.Handler.SubscriptionName;
+                var subscriptionId = subscriptionAndPerformanceConnectionStatsList.Handler.SubscriptionId;
+                var performance = subscriptionAndPerformanceConnectionStatsList.Performance;
+                
+                var connectionsAggregators = new List<SubscriptionConnectionStatsAggregator>(performance.Count);
+
+                // 1. get 'closed' connections info
+                while (performance.TryTake(out SubscriptionConnectionStatsAggregator stat))
+                {
+                    connectionsAggregators.Add(stat);
+                }
+                
+                // 2. get 'inProgress' connection info
+                using (Database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
```

Comment:
should we move this level up to avoid opening this inside for loop? @arekpalinski ?
same in  foreach (var kvp in _perSubscriptionBatchStats)

Same for read transactions: should we open them so many times?

Reply:
Let's call `AllocateOperationContext` outside the loop but `OpenReadTransaction` should be called as is - this way we have short running transactions

Reply:
done